src/awkward/_v2/contents/unionarray.py

```diff
@@ -1192,9 +1188,9 @@ def _to_numpy(self, allow_missing):
 
     def _completely_flatten(self, nplike, options):
         out = []
-        for content in self._contents:
+        for i in range(len(self._contents)):
             out.extend(
-                content[: self._tags.length]._completely_flatten(nplike, options)
+                self._contents[i][self._tags[i] :]._completely_flatten(nplike, options)
```

Comment:
This might be passing tests, but I think it's wrong. The number of items in the list `self._contents` is the number of types in the union, like two or three. The number of items in the array `self._tags` is the length of the union array itself, like a million. They can't both be indexed by the same `i`.

I think the original code you changed this from was wrong, too, so let's think it through. The purpose of this is to only descend the `_completely_flatten` on items in `self._contents[i]` that are reachable, i.e. to exclude the "hidden" data. After picking the single `Content` with `self._contents[i]`, the ones that are accessible are the ones that are picked out by index. The `self._index` includes all contents, not just `self._contents[i]`, so that, too, needs to be selected. The index values that you want correspond to `self._tags == i`, so I think the

```python
self._contents[i][self._tags[i]:]._completely_flatten(nplike, options)
```

has to be replaced by

```python
self._contents[i][self._index[self._tags == i]]._completely_flatten(nplike, options)
```

or maybe more directly

```python
self._contents[i]._carry(self._index._carry(self._tags.data == i, ...), ...)._completely_flatten(nplike, options)
```

with the appropriate arguments for `_carry` (to save it from having to go through all the user-friendliness checks that `__getitem__` does before it finds out that slicing by an array ought to be a call to `_carry`. Though `_carry` is somewhat restrictive: I think you have to give it an `Index`, rather than an array. Maybe you'd want to do this in several steps, getting arrays from `self._index` and `self._tags` and doing the slices in NumPy space before making an `Index` and carrying `self._contents[i]`.